File: benchmarks/7/runprocessing.go
package main

import (
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"os/signal"
	"runtime"
	"strconv"
	"strings"
	"syscall"
	"time"

	wp "./workerspool"
)

const (
	defaultMessageCount    = 100
	defaultInputStreamSize = 32
	defaultPoolSize        = 10
	gcEnvKey               = "GOGC"
)

var messageCount, poolSize, inputStreamSize int
var gcOff bool

type clientResult struct {
	count    int
	errCount int
	execTime float32
	inData   int
}

func init() {
	flag.IntVar(&messageCount, "message", defaultMessageCount, "Message count.")
	flag.IntVar(&poolSize, "pool", defaultPoolSize, "Pool size.")
	flag.IntVar(&inputStreamSize, "input", defaultInputStreamSize, "Count of input streams.")
	flag.Parse()
	gcOff = strings.ToLower(os.Getenv(gcEnvKey)) == "off"
	if gcOff {
		fmt.Println("GC is Off")
	}
}

func handler(worker int, data wp.ContentType) wp.ContentType {
	var answer string
	if value, err := strconv.Atoi(string(data)); err != nil {
		answer = fmt.Sprintf("error: %s", err)
	} else {
		// convert the number to hex format (base 16)
		answer = fmt.Sprintf("%x", value)
	}
	return wp.ContentType(answer)
}

func inputStream(index int, workers *wp.WorkerPool, resChan chan clientResult) {
	start := time.Now()
	result := clientResult{count: messageCount}
	for i := 0; i < messageCount; i++ {
		//simple random int value
		num := rand.Intn(1e6)
		numStr := fmt.Sprintf("%d", num)
		result.inData += len(numStr)
		if answer, err := workers.Exec(wp.ContentType(numStr)); err != nil {
			log.Println(err)
			result.errCount++
		} else {
			// return numbers in hex
			// check it
			resNum := string(*answer)
			if resNum != fmt.Sprintf("%x", num) {
				log.Printf("%d:%d %s -> %s\n", index, i, numStr, resNum)
				result.errCount++
			}
		}
	}
	result.execTime = float32(time.Since(start)) / 1E9
	resChan <- result
}

func showResult(resChan chan clientResult) {
	if gcOff {
		defer runtime.GC()
	}
	waitResult := true
	countRes := 0
	totalResCount := 0
	totalResTime := float32(0)
	totalData := 0
	waitTime := float32(0)
	start := time.Now()
	for waitResult {
		res := <-resChan
		if res.count > 0 {
			countRes++
			totalData += res.inData
			totalResTime += res.execTime
			totalResCount += res.count
			waitResult = countRes < inputStreamSize
			if !waitResult {
				waitTime = float32(time.Since(start)) / 1E9
			}
		} else {
			waitResult = false
		}
	}
	log.Printf(`
	input data size: %d
	msg count: %d
	avg worker time (ms): %.6f
	avg msg exec time (ms): %.6f
	client wait (sec): %.6f
	`,
		totalData,
		totalResCount*inputStreamSize,
		totalResTime*1000./float32(totalResCount),
		totalResTime*1000./float32(totalResCount*inputStreamSize),
		waitTime)
}

func main() {
	pool := wp.NewWorkerPool(poolSize, handler, 1)
	signalChannel := make(chan os.Signal, 1)
	signal.Notify(signalChannel, os.Interrupt)
	signal.Notify(signalChannel, syscall.SIGTERM)
	resChan := make(chan clientResult, inputStreamSize)
	defer close(resChan)
	for i := 0; i < inputStreamSize; i++ {
		go inputStream(i, pool, resChan)
	}
	go showResult(resChan)
	for msg := range signalChannel {
		log.Printf("out with: %s\n", msg)
		break
	}
	pool.Stop()
}

File: benchmarks/7/runprocessing.py

import asyncio
import functools
import random
import signal
import sys
import time

DEFAULT_INPUT_SIZE = 100
DEFAULT_MSG_COUNT = 10
DEFAULT_POOL_SIZE = 10
BUFFER_SIZE = 10000


async def client_input(
        index: int, msg_count: int, input_queue: asyncio.Queue):
    """Source data from clients.
    """
    start_time = time.time()
    total_data_size = 0
    total_msg_count = 0
    for msg_index in range(1, msg_count + 1):
        future = asyncio.Future()
        num = random.randint(0, 1000 ** 2)
        str_num = str(num)
        msg = {
            "id": msg_index,
            "future": future,
            "value": str_num,
        }
        total_data_size += sys.getsizeof(str_num)
        await input_queue.put(msg)
        while not future.done():
            await asyncio.sleep(0)

        res = future.result()
        if int(res["value"], 16) != num:
            print(
                "client {} result: {} source value: {}".format(
                    index, res, num))
        else:
            total_msg_count += 1

        await asyncio.sleep(0)

    return (
        total_msg_count,
        time.time() - start_time,
        total_data_size,
    )


async def worker(index: int, input_queue: asyncio.Queue):
    """Handler of worker.
    """
    print("worker {} started".format(index))
    active = True
    while active:
        msg = await input_queue.get()
        if msg.get("exit"):
            active = False
            print("worker {} stopped".format(index))
        else:
            future = msg["future"]
            del msg["future"]
            msg["worker"] = index
            msg["value"] = "{0:x}".format(int(msg["value"]))
            future.set_result(msg)

        await asyncio.sleep(0)


async def run_workers(
        input_queue: asyncio.Queue,
        pool: int,
        **kwargs):
    """Starting workers.
    """
    coroutines = [
        worker(i, input_queue)
        for i in range(1, pool + 1)
    ]
    completed, _ = await asyncio.wait(coroutines)
    return all(
        item.result()
        for item in completed)


async def run_clients(
        input_queue: asyncio.Queue,
        input: int,
        message: int,
        **kwargs):
    """Method for starting all clients.
    """
    coroutines = [
        client_input(i, message, input_queue)
        for i in range(input)
    ]
    start_time = time.time()
    completed, _ = await asyncio.wait(coroutines)
    total_data_size = 0
    total_exec_time = 0
    total_msg_count = 0
    for item in completed:
        msg_count, exec_time, data_size = item.result()
        total_data_size += data_size
        total_exec_time += exec_time
        total_msg_count += msg_count

    print(
        """
        input data size: {:d}
        msg count: {:d}
        avg client time (ms): {:.6f}
        avg msg exec time (ms): {:.6f}
        client wait (sec): {:.6f}""".format(
            total_data_size,
            total_msg_count,
            total_exec_time * 1000 / float(input),
            total_exec_time * 1000 / float(input * msg_count),
            time.time() - start_time))

#  #run#  #
run_data = {
    "input": DEFAULT_INPUT_SIZE,
    "message": DEFAULT_MSG_COUNT,
    "pool": DEFAULT_POOL_SIZE,
}

cmd_args = set(sys.argv[1:])
for key in run_data:
    for cmd_arg in cmd_args:
        try:
            arg, val = cmd_arg.split("=")
        except ValueError:
            continue
        else:
            if '-{}'.format(key) == arg:
                try:
                    run_data[key] = int(val)
                except (TypeError, ValueError):
                    continue


def ask_exit(sig_name: str, msg_queue: asyncio.Queue, workers: int):
    """Exit signal handler.
    """
    print("Exit signal:", sig_name)
    for _ in range(workers):
        msg_queue.put_nowait({"exit": True})


loop = asyncio.get_event_loop()
try:
    queue = asyncio.Queue(maxsize=BUFFER_SIZE)
    for signame in ('SIGINT', 'SIGTERM'):
        loop.add_signal_handler(
            getattr(signal, signame),
            functools.partial(
                ask_exit, signame, queue, run_data["pool"]))

    loop.run_until_complete(
        asyncio.gather(
            run_workers(input_queue=queue, **run_data),
            run_clients(input_queue=queue, **run_data)
        ))
finally:
    loop.close()
